### PR TITLE
feat(swarm): multi-agent swarm over the specialist sub-agent

### PR DESCRIPTION
**Pull request by gnanam1990**
## Summary

Adds **`internal/swarm`** — a multi-agent SWARM layered on top of zero's single sub-agent (`internal/specialist`). An orchestrator can spawn and coordinate **multiple** specialist members concurrently under a named team, have them communicate via per-agent mailboxes, hand work off, and adopt orphaned tasks.

Everything is **additive and opt-in**: the existing single `Task` tool is unchanged; the swarm tools only act when invoked.

## What's included

| Piece | File | Notes |
|---|---|---|
| Agent roster | `definition.go` | `Registry` + `Definition`; built-in `teammate`/`subagent`, plus user-defined agents via `Options.Definitions` |
| Mailbox | `mailbox.go` | per-team/per-agent JSON inbox, owner-only **0600**, file-locked (stale-break), **size-capped**, **path-confined**, malformed **fails closed** |
| Coordinator | `coordinator.go` | in-memory task registry with terminal-state guards + stable per-agent colors |
| Team roster | `team.go` | **bounded** concurrent members per team with a FIFO overflow **queue** (drains one-per-slot) |
| Lifecycle | `lifecycle.go` | spawn / handoff / orphan-adoption; **bounded relaunch** on temporary failures |
| Launcher seam | `member.go`, `launcher_specialist.go` | `MemberLauncher` interface (+`FuncLauncher`); production adapter over `specialist.Executor` |
| Tools | `tools.go` | `swarm_spawn`, `swarm_send`, `swarm_inbox`, `swarm_status`, `swarm_handoff`, `swarm_collect` |
| Wiring | `internal/cli/app.go`, `exec.go` | registers swarm tools alongside specialist tools; swarm lifetime paired with the specialist runtime |

## Security / fail-closed

- **Policy inheritance, never widened.** Each member is launched through `specialist.Executor` and inherits the orchestrator's **model + permission mode**; a definition that asks for a *more* permissive mode is clamped down to the parent's, and unknown modes rank strictest. Members run under the same sandbox/policy as the parent — the swarm never grants a member more authority.
- **Mailboxes are owner-only (0600)**, written atomically under an exclusive lock; names are sanitized to a single safe path segment and the resolved path is verified to stay under the base dir.
- **Caps + fail-closed**: per-message and per-inbox size caps reject oversize input; a full inbox is rejected; malformed inbox JSON surfaces an error rather than being silently reset; the team slot cap queues overflow rather than over-spawning.
- Mailbox state lives under the workspace (`.zero/swarm`) so its writes fall within the sandbox write rules.
- Members run under the Swarm's **base context** (not the per-call tool context) so a background member is not cancelled the instant `swarm_spawn` returns; `Swarm.Close` cancels them all on shutdown.

## Tests

Per-layer coverage incl. `-race`: roster lookup/override, mailbox (sanitize, oversize, full, malformed-fails-closed, path confinement, 0600 mode, concurrent sends), coordinator (terminal guards, reassign, color stability, concurrent registers), team (admit/queue, permission clamp, model resolve), lifecycle (completion, **concurrency cap + queue drain**, **retry/exhaustion**, permanent-error no-retry, **handoff**, **orphan-adoption**), and all six tools through the registry dispatch (registration, policy/cwd threading, validation, permission gating).

## Gates

`gofmt` clean · `go vet` clean · host + `GOOS=linux` build clean · full `go test ./...` pass · `staticcheck` no new findings · `govulncheck` 0 (code affected by 0) · `deadcode` no new entries vs `main`.

## Deferred (follow-ups)

- **Scheduling** (cron/wakeup/kairos) — out of scope for this PR.
- **Daemon-pool-backed members** — the launcher seam supports it; bridging the daemon worker protocol to specialist result capture is its own change.

## Summary by CodeRabbit

* **New Features**
  * Added an in-memory multi-agent swarm with task coordination, status tracking, reassignment, handoff, orphan adoption, and team-scoped collection.
  * Introduced a persistent mailbox with safe path confinement, size/cap limits, and locked JSON inbox delivery/consumption.
  * Added swarm CLI tools to spawn tasks, send/read inbox messages, check status, hand off work, and collect results.
* **Bug Fixes**
  * Specialist execution autonomy now follows the parent permission mode.
  * Member panics are recovered and returned as task errors.
* **Tests**
  * Added unit and integration coverage for coordinator, lifecycle, tools, mailbox, and concurrency/edge cases.

### CODE REVIEW

File: internal/swarm/coordinator.go [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+	if !ok {
+		return fmt.Errorf("%w: %s", ErrUnknownTask, id)
+	}
+	if t.Status.terminal() && t.Status != status {
+		return fmt.Errorf("swarm: task %s is %s (terminal); cannot move to %s", id, t.Status, status)
+	}
+	t.Status = status
+	t.UpdatedAt = c.now()
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (gnanam1990) [reply internal/swarm/coordinator.go:131]: Fixed in 192f93e: `SetStatus` now rejects unknown `TaskStatus` values via `TaskStatus.valid()` before any mutation (fail closed). Test: `TestCoordinatorRejectsInvalidStatus`.

File: internal/swarm/launcher_specialist.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+			Description: spec.AgentType + " · team " + spec.Team,
+		}, specialist.TaskRunOptions{
+			ParentModel: spec.Model,
+			Cwd:         spec.Cwd,
+		})
+		if err != nil {
+			return MemberResult{}, err
+		}
+		return MemberResult{Result: res.Result.Output, SessionID: res.SessionID}, nil
+	}}
```
[the remark this answers is not in the document]
Comment (gnanam1990) [reply internal/swarm/launcher_specialist.go:46]: Fixed in 192f93e. `PermissionMode` is now threaded through `specialist.TaskRunOptions` → `BuildArgs`/`BuildResumeArgs`. A headless child runs `--auto high` (unsafe) ONLY when the parent is unsafe or no mode is supplied (preserves the existing Task tool behavior); any other parent mode yields non-unsafe `--auto low`, so a member never gains more authority than its orchestrator. Also corrected the swarm permission vocabulary to the real runtime modes (ask/spec-draft/auto/unsafe) and made the launcher substitute `auto` for an empty mode. Tests: `TestBuildArgsAutonomyHonorsPermissionMode`, `TestBuildResumeArgsAutonomyHonorsPermissionMode`, `TestSpecialistAutonomyByPermissionMode`.

File: internal/swarm/mailbox.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+// and verifies it cannot escape BaseDir (defense in depth on top of sanitize).
+func (m *Mailbox) inboxPath(team, agent string) (string, error) {
+	dir := filepath.Join(m.BaseDir, sanitizeName(team), "inboxes")
+	path := filepath.Join(dir, sanitizeName(agent)+".json")
+	// Confinement check: the resolved path must stay under BaseDir.
+	rel, err := filepath.Rel(m.BaseDir, path)
+	if err != nil || rel == ".." || strings.HasPrefix(rel, ".."+string(os.PathSeparator)) {
+		return "", fmt.Errorf("swarm: inbox path escapes base dir (team=%q agent=%q)", team, agent)
+	}
+	return path, nil
```
[the remark this answers is not in the document]
Comment (gnanam1990) [reply internal/swarm/mailbox.go:127]: Fixed in 192f93e: `ensureInboxDir` resolves symlinks (`EvalSymlinks`) and rejects an inbox dir whose real path escapes the real BaseDir. Test: `TestMailboxRejectsSymlinkedInboxDir`.

File: internal/swarm/mailbox.go [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+	encoded, err := json.Marshal(msg)
+	if err != nil {
+		return fmt.Errorf("swarm: encode message: %w", err)
+	}
+	if len(encoded) > m.maxMessageBytes() {
+		return fmt.Errorf("%w: %d > %d bytes", ErrMessageTooLarge, len(encoded), m.maxMessageBytes())
+	}
+	if err := os.MkdirAll(filepath.Dir(path), inboxDirMode); err != nil {
+		return fmt.Errorf("swarm: create inbox dir: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (gnanam1990) [reply internal/swarm/mailbox.go]: Fixed in 192f93e: `ensureInboxDir` chmods the team + inboxes dirs to 0700 even when they pre-existed with broad perms (after the symlink-confinement check, so a chmod can never land outside base). Test: `TestMailboxTightensDirPerms`.

File: internal/swarm/mailbox.go [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+		}
+		if !errors.Is(err, os.ErrExist) {
+			return nil, fmt.Errorf("swarm: acquire lock: %w", err)
+		}
+		// Lock held: break it if stale, otherwise wait.
+		if info, statErr := os.Stat(lockPath); statErr == nil {
+			if time.Since(info.ModTime()) > lockStaleAfter {
+				os.Remove(lockPath)
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (gnanam1990) [reply internal/swarm/mailbox.go:375]: Fixed in 192f93e: each holder writes a unique token into the lock file and release is ownership-aware — it removes the lock only if it still holds OUR token, so a stale-broken writer cannot delete a newer writer’s lock. Stale-break is also content-conditional. Test: `TestMailboxLockReleaseIsOwnershipAware`.

File: internal/swarm/member.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+// Launch starts the RunFunc for spec and returns a handle to await it.
+func (l FuncLauncher) Launch(ctx context.Context, spec MemberSpec) (MemberHandle, error) {
+	if l.Run == nil {
+		return nil, errors.New("swarm: FuncLauncher requires a Run func")
+	}
+	h := &funcHandle{id: spec.ID, done: make(chan struct{})}
+	go func() {
+		defer close(h.done)
+		h.res, h.err = l.Run(ctx, spec)
+	}()
```
[the remark this answers is not in the document]
Comment (gnanam1990) [reply internal/swarm/member.go:80]: Fixed in 192f93e: `FuncLauncher` recovers panics inside the member goroutine and surfaces them as that member’s error, so a panic fails one member instead of crashing the orchestrator. Test: `TestFuncLauncherRecoversPanic`.

File: internal/swarm/tools.go [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+// status/collect output.
+func collapse(s string) string {
+	s = strings.ReplaceAll(s, "\n", " ")
+	s = strings.Join(strings.Fields(s), " ")
+	const max = 200
+	if len(s) > max {
+		return s[:max] + "…"
+	}
+	return s
+}
```
[the remark this answers is not in the document]
Comment (gnanam1990) [reply internal/swarm/tools.go:402]: Fixed in 192f93e: `collapse` truncates on rune boundaries (`[]rune` slice), never mid-character. Test: `TestCollapseRuneSafeTruncation`.

**Comment by gnanam1990** [comment]
@coderabbitai review

All 7 findings addressed in 192f93e (replies on each thread):
1. **SetStatus** — rejects unknown statuses (fail closed).
2. **PermissionMode enforcement (critical)** — now threaded through `specialist.TaskRunOptions` into the subprocess; a member runs `--auto high` only when the parent is unsafe (else non-unsafe `--auto low`), so authority is never widened. Existing Task tool behavior unchanged (empty mode ⇒ `--auto high`).
3. **Symlink confinement** — `EvalSymlinks` rejects an inbox dir escaping BaseDir.
4. **Dir perms** — team + inboxes dirs tightened to 0700 even if pre-existing.
5. **Lock split-brain** — ownership-token-aware release.
6. **Member panic** — recovered per-member, never crashes the orchestrator.
7. **collapse UTF-8** — rune-boundary truncation.

Tests added for each. Gates: gofmt · vet · build (host+linux) · test -race · staticcheck (no new) · govulncheck (0) · deadcode (no new) — all green.

**Review by Vasanthdev2004 — CHANGES_REQUESTED** [review]
## Review of PR #207 — feat(swarm): multi-agent swarm over the specialist sub-agent

Reviewed the 18 changed files against merge base `ba8fbb8b`. Locally I ran:
- `go test -count=1 ./internal/swarm/... ./internal/specialist/...` — passes on a single run
- `go test ./...` — full tree clean
- `go vet ./internal/swarm/... ./internal/specialist/...` — clean
- `go build ./...` — clean
- `gofmt -l` — clean
- 30× re-runs of `TestMailboxConcurrentSends` on Windows — **4/30 failed** (see blocker below)

The package is well structured: tight interfaces (`MemberLauncher`, `MemberHandle`), policy clamping that never widens authority, owner-only 0600 mailboxes, size caps that fail closed, path-confinement checks, orphan-adoption, bounded restarts. The documentation mirrors the JS reference cleanly while hardening it.

### Blocker — Windows mailbox lock contention drops messages (mailbox.go: `acquireLock`)

`TestMailboxConcurrentSends` is flaky. Over 30 runs on Windows I saw **4 failures** (got 28/29/29 instead of 30). The error logged from the failing goroutine is:

```
swarm: acquire lock: open .../bob.json.lock: Access is denied.
```

`acquireLock` only treats `os.ErrExist` as "lock held, retry"; anything else is a hard error. On Windows the failure path is `ERROR_SHARING_VIOLATION` (surfaced as `Access is denied`) because the previous holder's `release()` opens the lock file with a read handle via `os.ReadFile` to check the ownership token *while another goroutine is racing to acquire*. That handle blocks a concurrent `O_EXCL` create with `ERROR_ACCESS_DENIED`, not `ErrExist`, so the message is dropped.

The pattern is exactly the one called out in the existing `release` comment about split-brain writers — but on Windows, even the "normal" read-then-remove pattern can transiently fail on a sibling acquisition.

**Suggested fix** (any one of these):
1. Treat `syscall.ERROR_ACCESS_DENIED` and `syscall.ERROR_SHARING_VIOLATION` (Windows-only) as "lock contended, retry" alongside `os.ErrExist`. One-line change in the error check at `mailbox.go:360`.
2. Avoid opening the lock file for read during release: encode the token in the *filename* (e.g. `lockPath = <base>.lock.<token>`) so release can `os.Remove` directly without a read.
3. On Windows, use `LockFileEx` from `golang.org/x/sys/windows` for true OS-level locking instead of relying on `O_EXCL` semantics that don't fully cover sharing violations.

I'd lean toward (1) because it's the smallest change and the race window is small. (2) is cleaner long-term.

A test that exercises this on Windows (the existing `TestMailboxConcurrentSends` already does, but is the only one) — could be made deterministic by raising the goroutine count to e.g. 200 with a 10s lock timeout so the failure is observed >50% of the time on Windows, then asserting all 200 land. That would also prevent a future regression that re-introduces a similar race.

### Non-blocking notes

1. **mailbox.go:17** — comment says it "mirrors reference-swarm-code-agent-js/communication/mailbox.js". The PR description doesn't mention it. Confirm clean-room only (no copy-paste from a third-party reference project). Same for the `agentColors` palette, `builtinDefinitions` (the doc comment at `definition.go:108` says it mirrors `definitions/swarm.js`), and the lifecycle/supervisor/team layout. Skim those doc-comments are accurate; the Go code is the team's own.
2. **mailbox.go:280** — `maxFile := m.maxMessages() * m.maxMessageBytes() * 2` — the `* 2` is to cover pretty-printed JSON vs compact. That's a magic constant with no comment; a one-liner explaining the bound would help the next reader.
3. **mailbox.go:215** — `Send` rewrites the whole inbox on every send (read all → append one → write all). Fine at 1000 msgs / 64 KiB cap, but at the cap it allocates ~64 MiB to write a 64-byte message. Consider documenting the O(N) per-send cost, or moving to append-only with periodic compaction.
4. **team.go:209** — `permissionRank` ranks spec-draft between ask and auto. Is that the order you want? `spec-draft` shouldn't widen authority at all (it restricts what the child can do), so it might deserve rank 0 alongside unknown. Worth confirming the intent.
5. **lifecycle.go:60-65** — bounded relaunch uses `m.restarts < maxMemberRestarts` to gate the retry, but the counter is per-member, not per-spec. After a relaunch, the *new* handle is `nh` and `m.handle = nh` — that's fine — but if a *different* spec is on the same member struct… actually that doesn't happen because `Member.ID == MemberSpec.ID`. Confirmed safe, but a one-line comment about the per-member-id binding would help.
6. **launcher_specialist.go:30-32** — the empty-PermissionMode fallback to `"auto"` is a smart hardening (good!). Worth a comment that the swarm overrides the historical "empty ⇒ high" Task-tool default, so reviewers don't try to "simplify" it.
7. **tools.go:296-299** — `handoffTool.Safety` is `SideEffectShell / PermissionPrompt` because it spawns a member. Correct, but consider also classifying the optional mailbox write as `SideEffectWrite` in the `Reason` text so the user sees both effects in the permission prompt.

### Verdict

**Request changes** on the Windows mailbox lock bug only. Everything else is non-blocking.

Once the mailbox `acquireLock` treats `ERROR_ACCESS_DENIED` / `ERROR_SHARING_VIOLATION` as a retry signal (and ideally adds a higher-fanout Windows test), this is good to ship.

**Comment by gnanam1990** [comment]
@Vasanthdev2004 thanks for the thorough review — all addressed in 2a2c6a5.

**Blocking — Windows mailbox lock:** `acquireLock` now treats `ERROR_ACCESS_DENIED` (5) and `ERROR_SHARING_VIOLATION` (32) as lock-contended (retry/wait) instead of a fatal error, via a platform-aware `isLockContended` split across `lock_windows.go` / `lock_other.go` (POSIX keeps the `os.ErrExist`-only check). `TestMailboxConcurrentSends` is raised to **200 goroutines / 10s lock timeout** and now asserts **zero** send failures, so a regression that re-breaks contention handling fails the test. Verified it cross-builds: `GOOS=windows go build ./internal/swarm/`.

**Non-blocking notes:**
1. Removed the `reference-…js` / third-party project names from doc comments (clean-room; the Go is original).
2. Documented the `×2` inbox-size bound (pretty-print overhead).
3. Documented `Send`'s O(N) rewrite cost + the append-only fallback if caps grow.
4. `permissionRank`: `spec-draft` now ranks **below** `ask` (it never widens authority).
5. Documented `watch`'s per-member-id retry binding (`Member.ID == MemberSpec.ID`).
6. Launcher already documents the empty⇒`auto` override; kept.
7. `handoff` `Safety.Reason` now also mentions the inbox write.

Gates: gofmt · vet · build (host+linux+**windows**) · test `-race` · staticcheck (no new) · govulncheck (0) · deadcode (no new) — all green.

**Review by Vasanthdev2004 — APPROVED** [review]
## Re-review of PR #207 round 2 — commit 2a2c6a5

Verified the round-2 fix on commit `2a2c6a5` (10 new files vs main, +3182/-19 lines; +93/-29 vs the prior commit `192f93e`).

### Blocker addressed — Windows mailbox lock

`acquireLock` in `mailbox.go:365` now delegates to a platform-aware `isLockContended(err)`:

- `internal/swarm/lock_windows.go` (new, `//go:build windows`) — matches `os.ErrExist` **or** `syscall.ERROR_ACCESS_DENIED` (5) **or** `syscall.ERROR_SHARING_VIOLATION` (32). Constants are commented with their semantics and the rationale ("concurrently open or pending deletion").
- `internal/swarm/lock_other.go` (new, `//go:build !windows`) — POSIX-only `os.ErrExist`. Behavior is unchanged on Linux/macOS.

The implementation is the right shape: it isolates the platform difference in two small files with one function each, keeps the cross-platform `acquireLock` call site a one-line change, and uses `errors.Is` so wrappers around the syscall errors would still match. No performance impact on the happy path.

### Regression test made deterministic

`TestMailboxConcurrentSends` (mailbox_test.go:247) now:
- Runs **200 goroutines** (was 30).
- Uses a **10s LockTimeout** (was default 5s).
- Asserts **0 failures** via `atomic.Int32` — `t.Fatalf("%d concurrent Sends failed (lock contention not handled)", f)`.
- Still asserts all 200 messages landed.

I ran the new test 30 times on this Windows box: **30/30 pass, 0 fail**. Previously it was 26/30 (4 lost messages with the "Access is denied" error). The fix holds.

### All 7 non-blocking notes addressed

1. ✅ Doc comments no longer reference the third-party JS project — `mailbox.go:15-16`, `coordinator.go:62`, `definition.go:21/26/46/101` all stripped of the "mirrors X.js" wording.
2. ✅ `* 2` pretty-print overhead in the inbox size bound is now commented (mailbox.go:281-285).
3. ✅ `Send`'s O(N) cost is documented with a switch-to-append-only guidance (mailbox.go:174-179).
4. ✅ `permissionRank` re-ordered: `spec-draft` now ranks **1**, `ask` **2**, with a comment explaining the spec-draft-never-widens-authority reasoning (team.go:220-227).
5. ✅ `watch` per-member-id retry binding documented (lifecycle.go:56-61).
6. ✅ `handoffTool.Safety().Reason` now mentions the optional inbox write (tools.go:298).
7. ➖ `launcher_specialist.go` empty-`PermissionMode` fallback — already documented in the original PR; no change needed.

### Local gates

- `go test -count=1 ./internal/swarm/... ./internal/specialist/...` — **pass**
- `go vet ./...` — **clean**
- `go build ./...` — **clean**
- `gofmt -l` — **clean**
- Full `go test ./...` — only one failure: `TestRunExecUsesProjectConfigAndOpenAICompatibleProvider` in `internal/cli`. I checked out the merge base (`ba8fbb8`) and confirmed this same test fails identically on `main` with `[zero] anthropic provider anthropic requires official baseURL https://api.anthropic.com` — it's an environment-dependent test, **unrelated to this PR**. CI (Linux/macOS/Windows) passes it, so it must be a sandbox-network policy difference in this local runner.
- CI on the PR — all 7 checks green (Zero Review, ubuntu/macOS/Windows smoke, performance smoke, security & code health, CodeRabbit).

### Verdict

**Approve.** All blocking and non-blocking items from my previous review are addressed, the test is now a deterministic Windows regression guard, and the fix is minimal, well-commented, and platform-isolated. Good to merge.
